packages/network-controller/src/provider-api-tests/shared-tests.ts

```diff
@@ -0,0 +1,2673 @@
+
+        // { name: 'eth_sendRawTransaction', numberOfParameters: 1 },
+        // { name: 'eth_sendTransaction', numberOfParameters: 1 },
+        // { name: 'eth_sign', numberOfParameters: 2 },
+
+        { name: 'eth_createAccessList', numberOfParameters: 2 },
+        { name: 'eth_getLogs', numberOfParameters: 1 },
+        { name: 'eth_getProof', numberOfParameters: 3 },
+        { name: 'eth_getWork', numberOfParameters: 0 },
+        { name: 'eth_maxPriorityFeePerGas', numberOfParameters: 0 },
+        { name: 'eth_submitHashRate', numberOfParameters: 2 },
+        { name: 'eth_submitWork', numberOfParameters: 3 },
+        { name: 'eth_syncing', numberOfParameters: 0 },
+        { name: 'eth_feeHistory', numberOfParameters: 3 },
+        { name: 'debug_getRawHeader', numberOfParameters: 1 },
+        { name: 'debug_getRawBlock', numberOfParameters: 1 },
+        { name: 'debug_getRawTransaction', numberOfParameters: 1 },
+        { name: 'debug_getRawReceipts', numberOfParameters: 1 },
+        { name: 'debug_getBadBlocks', numberOfParameters: 0 },
+      ];
+      notHandledByMiddleware.forEach(({ name, numberOfParameters }) => {
+        describe(`method name: ${name}`, () => {
+          testsForRpcMethodNotHandledByMiddleware(name, {
+            providerType,
+            numberOfParameters,
+          });
+        });
+      });
+    });
+
+    describe('methods handled by middleware that returns static a result', () => {
+      const notHandledByMiddleware = [
+        { name: 'eth_accounts', numberOfParameters: 0, result: [] },
+        { name: 'eth_coinbase', numberOfParameters: 0, result: null },
+        { name: 'eth_hashrate', numberOfParameters: 0, result: '0x00' },
+        { name: 'eth_mining', numberOfParameters: 0, result: false },
+      ];
+      notHandledByMiddleware.forEach(({ name, numberOfParameters, result }) => {
+        describe(`method name: ${name}`, () => {
+          testsForRpcMethodWithStaticResult(name, {
+            providerType,
+            numberOfParameters,
+            result,
+          });
+        });
+      });
+    });
+
+    describe('methods with block hashes in their result', () => {
+      const methodsWithBlockHashInResponse = [
+        { method: 'eth_getTransactionByHash', numberOfParameters: 1 },
+        { method: 'eth_getTransactionReceipt', numberOfParameters: 1 },
+      ];
+
+      methodsWithBlockHashInResponse.forEach(
+        ({ method, numberOfParameters }) => {
+          describe(`method name: ${method}`, () => {
+            testsForRpcMethodsThatCheckForBlockHashInResponse(method, {
+              providerType,
+              numberOfParameters,
+            });
+          });
+        },
+      );
+    });
+
+    describe('methods that assume there is no block param', () => {
+      const assumingNoBlockParam = [
+        // { name: 'eth_getFilterLogs', numberOfParameters: 1 },
+        // { name: 'eth_blockNumber', numberOfParameters: 0 },
```

Comment:
Why does this one not work?

Reply:
The mocks were getting mangled because they conflict with the mocks for mockNextBlockTrackerRequest. The number of times eth_blockNumber will be called, and whether the caller is the block tracker, or this particular request, was very hard to determine, and to make it work in the context of `no-block-param` tests was a losing battle. When we mock eth_blockNumber to return an error, both this call get retried, as well as the blockTracker requests, and the resulting number of calls that will happen ends up being based on the number of times you advance the clock. Since web3-provider-engine uses additional setTimeouts (in stoplight, for example), we have to advance the clock after making each request. When you combine all these things, it makes this method incredibly difficult to test with these shared tests.

Perhaps now that the rest of the tests are done, we can revisit this one and write some tests specifically to handle it. The tests that fail for eth_blockNumber are only the ones relating to retries / failures, since those cause a large number of requests to happen. The others pass (cache hits/miss tests, etc). We could guard the retry tests in a if (method !== 'eth_blockNumber'), and then uncomment this line, then write a whole new test helper setup that will work for this method (mocking eth_blockNumber & advancing the clock in a way that will work for this case).